lib/src/transformers.dart

```diff
@@ -133,145 +133,155 @@ dynamic convertColumn(
-  switch (stringValue) {
+bool? toBoolean(dynamic value) {
+  switch (value) {
     case 't':
       return true;
     case 'f':
       return false;
     default:
+      if (value is bool) {
+        return value;
+      }
       return null;
   }
 }
 
-DateTime toDate(String stringValue) {
-  return DateTime.parse(stringValue);
+double? toDouble(dynamic value) {
+  if (value is String) {
+    try {
+      return double.parse(value);
+    } catch (_) {}
+  }
+  if (value is double) {
+    return value;
+  }
+  return null;
 }
 
-List<DateTime> toDateRange(String stringValue) {
-  final arr = json.decode(stringValue);
-  return [DateTime.parse(arr[0] as String), DateTime.parse(arr[1] as String)];
+int? toInt(dynamic value) {
+  if (value is String) {
+    try {
+      return int.parse(value);
+    } catch (_) {}
+  }
+  if (value is int) {
+    return value;
+  }
+  return null;
 }
 
-double toDouble(String stringValue) {
-  return double.parse(stringValue);
+dynamic toJson(dynamic value) {
+  if (value is String) {
+    try {
+      return json.decode(value);
+    } catch (error) {
+      print('JSON parse error: $error');
+      return value;
+    }
+  }
+  return value;
 }
 
-int toInt(String stringValue) {
-  return int.parse(stringValue);
-}
+/// Converts a Postgres Array into a native Dart array
+///
+///``` dart
+/// @example toArray('{"[2021-01-01,2021-12-31)","(2021-01-01,2021-12-32]"}', 'daterange')
+/// //=> ['[2021-01-01,2021-12-31)', '(2021-01-01,2021-12-32]']
+/// @example toArray([1,2,3,4], 'int4')
+/// //=> [1,2,3,4]
+///  ```
+dynamic toArray(dynamic value, String type) {
+  if (value is! String) {
+    return value;
+  }
+
+  // trim Postgres array curly brackets
+  final lastIdx = value.length - 1;
+  final closeBrace = value[lastIdx];
+  final openBrace = value[0];
 
-List<int> toIntRange(String stringValue) {
-  final arr = json.decode(stringValue);
-  return [int.parse(arr[0] as String), int.parse(arr[1] as String)];
+  // Confirm value is a Postgres array by checking curly brackets
+  if (openBrace == '{' && closeBrace == '}') {
+    final valTrim = value.substring(1, lastIdx);
+    List arr;
+
+    // TODO: find a better solution to separate Postgres array data
+    try {
+      arr = json.decode('[$valTrim]') as List;
+    } catch (_) {
+      // WARNING: splitting on comma does not cover all edge cases
+      arr = valTrim != '' ? valTrim.split(',') : [];
+    }
+
+    return arr.map((val) => convertCell(type, val)).toList();
+  }
+
+  return value;
 }
 
-dynamic toJson(String stringValue) {
-  return json.decode(stringValue);
+/// Fixes timestamp to be ISO-8601. Swaps the space between the date and time for a 'T'
+/// See https://github.com/supabase/supabase/issues/18
+///
+///```dart
+/// @example toTimestampString('2019-09-10 00:00:00')
+/// => '2019-09-10T00:00:00'
+/// ```
+String? toTimestampString(dynamic value) {
+  if (value is String) {
+    return value.replaceAll(' ', 'T');
+  }
+  return null;
 }
```

Comment:
@bdlukaa I changed the return type to `String?` so that it will work when the `value` is actually null, but what do you think about this implementation?

Reply:
LGTM! But why does value have to be dynamic?

Reply:
It's set to `dynamic`, because the type of `value` passed onto `convertCell()` on line 136 is `dynamic`. Do you think we should accept `String?` as the argument type of `value` of `toTimestampString()` and type cast it at line 162 like this?

```dart
return toTimestampString(value as String?); 
```

Reply:
Maybe `value.toString`?

Reply:
@bdlukaa That's a great idea! I will do that!